### DISCUSSION TITLE
Support Grouping on Table Layout

### DISCUSSION THREAD

**Discussion by bryantgillespie** [directus/directus#20318] [Feature Requests]
### Summary

Within the Directus app, on the Table Layout, add support for grouping items together visually by a user-defined field.

### Motivation

This is one of the killer features of apps like Notion and Airtable that I love and use all the time. Past clients loved it as well. 

It makes parsing and making sense of the data in the table layout much easier for users. And makes it easier to update / manage data as well.

### Detailed Design

![ScreenShot 2023-11-03 at 09 51 38@2x](https://github.com/directus/directus/assets/23302570/ce5f8609-86d6-4d62-83e7-f16308d4857b)

![table-layout-grouping](https://github.com/directus/directus/assets/23302570/283158df-140f-4649-99e2-0b95d673908e)

![ScreenShot 2023-11-03 at 09 56 14@2x](https://github.com/directus/directus/assets/23302570/c7cd803a-4239-474a-8833-1913ea89d7ff)

## Examples

**Airtable**

Airtable supports multiple groupings up to 3 levels - each is collapsible.

![ScreenShot 2023-11-03 at 08 23 13@2x](https://github.com/directus/directus/assets/23302570/3ec54f12-bdfc-4281-ab35-38bffd4842ef)

Values for the group by field are updated on drag and drop.

https://github.com/directus/directus/assets/23302570/c48666ce-5dc7-4144-848a-d7fb9e8996af

**Notion**

Notion supports only a single grouping.

Values for the group by field are also updated on drag and drop.

![ScreenShot 2023-11-03 at 09 48 39@2x](https://github.com/directus/directus/assets/23302570/de78f662-3c06-4b3f-9f17-151724387f1c)

![ScreenShot 2023-11-03 at 09 49 53@2x](https://github.com/directus/directus/assets/23302570/f35e7a92-6dcb-4f84-8339-3e1d2da983b2)

### Must haves:

- [ ]  Ability to group by root level fields on a collection
- [ ]  Expand / collapse groups
- [ ]  Respect manual sorting within groups

### Should haves:

- [ ]  Ability to group by relational fields
    - Would probably need to have some type of cap on how deep you can go though.

### Could haves:

- [ ]  Support grouping multiple times (nested grouping may be a better term?)

For example - assume we have three fields Type, Category, and Phase. Users may want to group by Type, then Category, then Phase
    - Type
        - Category
            - Phase
- [ ]  If Grouped By field is of the appropriate type - update the value of that field on drag and drop. 

Similar to how the Kanban layout functions. If I drag an item to a different list - the value of the group by field for that item is updated.

### Drawbacks

Nothing is ever as easy as it seems on the surface so I'm sure it's no picnic to implement this functionality but I don't see any drawbacks from adding it. 

### Alternatives

Directus doesn't support grouping currently so the alternative is just basically using the built-in sorting but that doesn't achieve quite the same result. 

We also only allow one sort field at a time within the app.

### Adoption Strategy

Could be a breaking change if there are API changes necessary to make this work. But otherwise it would just be added to the Directus app.

### Unresolved Questions

**Rendering performance**

I know with a large amount of data the Airtable and Notion tend to bog down a bit. Most likely because the browser is doing the heavy lifting of the grouping computation.

Directus is made to handle any amount of data whereas Airtable and Notion tend to be for smaller numbers of records and data. But we wouldn’t want user experience to suffer because of the grouping.

Maybe some type of [list virtualization](https://tanstack.com/virtual/v3) is required here (if we’re not using this pattern already).

**Pagination**

How would grouping work with pagination? It doesn’t seem like a great user experience to paginate in the middle of a group but given that you could have 50 rows or 5,000,000 in one collection - I’m not sure the exact solution here. 

Maybe this is where there could be changes to the APIs involved if any.

**Comment by jonathan-james-dev** [comment]
Would sort by multiple columns be out of the scope of this, similar to how Notion does it. I can see the two being closely related. Two birds with one stone :) 

![image](https://github.com/directus/directus/assets/131263649/f125b80f-f7ef-464e-a0c4-fb84ce9c9fb9)

**Comment by liamb0t** [comment]
Any updates on this?

**Reply by bryantgillespie** [reply]
No updates on this at the moment. I still believe it would be a valuable addition. 
But the scope is quite a bit larger than I had originally anticipated as it would require potentially breaking changes to the API as well to add full support. We have group_by query parameter now - but it's only used for aggregation.
